File: backend/src/features/users/users.controller.ts
import { Request, Response } from "express";
import * as usersService from "./users.service";
import { getRelationshipStatus } from "../relationships/relationship.service";
import { canViewProfile } from "../profiles/profiles.permissions";
import { getTopRecommendations } from "../recommendations/recommendations.service";

async function hasAccess(viewerId: string, targetId: string): Promise<boolean> {
  if (viewerId === targetId) return true;
  const status = await getRelationshipStatus(viewerId, targetId);
  if (canViewProfile(status)) return true;
  const recs = await getTopRecommendations(viewerId);
  return recs.includes(targetId);
}

export const getMe = async (req: Request, res: Response) => {
  try {
    const user = await usersService.getPublicUser(req.userId!);
    return res.json(user);
  } catch {
    return res.status(404).json({ message: "User not found" });
  }
};

export const getMyProfileData = async (req: Request, res: Response) => {
  try {
    const data = await usersService.getUserProfileData(req.userId!);
    return res.json(data);
  } catch {
    return res.status(404).json({ message: "User not found" });
  }
};

export const getMyBioData = async (req: Request, res: Response) => {
  try {
    const data = await usersService.getUserBioData(req.userId!);
    return res.json(data);
  } catch {
    return res.status(404).json({ message: "User not found" });
  }
};

export const getPublicUserById = async (req: Request, res: Response) => {
  try {
    const id = req.params.id as string;
    if (!await hasAccess(req.userId!, id)) return res.status(404).json({ message: "User not found" });
    const user = await usersService.getPublicUser(id);
    return res.json(user);
  } catch {
    return res.status(404).json({ message: "User not found" });
  }
};

export const getUserProfileById = async (req: Request, res: Response) => {
  try {
    const id = req.params.id as string;
    if (!await hasAccess(req.userId!, id)) return res.status(404).json({ message: "User not found" });
    const data = await usersService.getUserProfileData(id);
    return res.json(data);
  } catch {
    return res.status(404).json({ message: "User not found" });
  }
};

export const getUserBioById = async (req: Request, res: Response) => {
  try {
    const id = req.params.id as string;
    if (!await hasAccess(req.userId!, id)) return res.status(404).json({ message: "User not found" });
    const data = await usersService.getUserBioData(id);
    return res.json(data);
  } catch {
    return res.status(404).json({ message: "User not found" });
  }
};

File: backend-java/src/main/java/com/nexus/features/users/UsersController.java
package com.nexus.features.users;

import com.nexus.features.recommendations.RecommendationService;
import com.nexus.features.relationships.RelationshipService;
import com.nexus.util.AuthUtil;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.UUID;

@RestController
public class UsersController {

    private final UsersService usersService;
    private final RelationshipService relationshipService;
    private final RecommendationService recommendationService;

    public UsersController(UsersService usersService,
                           RelationshipService relationshipService,
                           RecommendationService recommendationService) {
        this.usersService = usersService;
        this.relationshipService = relationshipService;
        this.recommendationService = recommendationService;
    }

    // -------------------------------------------------------------------------
    // /me shortcuts
    // -------------------------------------------------------------------------

    @GetMapping("/me")
    public ResponseEntity<?> getMe() {
        try {
            return ResponseEntity.ok(usersService.getPublicUser(AuthUtil.getCurrentUserId()));
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("message", "User not found"));
        }
    }

    @GetMapping("/me/profile")
    public ResponseEntity<?> getMyProfileData() {
        try {
            return ResponseEntity.ok(usersService.getUserProfileData(AuthUtil.getCurrentUserId()));
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("message", "User not found"));
        }
    }

    @GetMapping("/me/bio")
    public ResponseEntity<?> getMyBioData() {
        try {
            return ResponseEntity.ok(usersService.getUserBioData(AuthUtil.getCurrentUserId()));
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("message", "User not found"));
        }
    }

    // -------------------------------------------------------------------------
    // /users/:id endpoints
    // -------------------------------------------------------------------------

    @GetMapping("/users/{id}")
    public ResponseEntity<?> getPublicUserById(@PathVariable UUID id) {
        try {
            if (!hasAccess(AuthUtil.getCurrentUserId(), id)) {
                return ResponseEntity.status(404).body(Map.of("message", "User not found"));
            }
            return ResponseEntity.ok(usersService.getPublicUser(id));
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("message", "User not found"));
        }
    }

    @GetMapping("/users/{id}/profile")
    public ResponseEntity<?> getUserProfileById(@PathVariable UUID id) {
        try {
            if (!hasAccess(AuthUtil.getCurrentUserId(), id)) {
                return ResponseEntity.status(404).body(Map.of("message", "User not found"));
            }
            return ResponseEntity.ok(usersService.getUserProfileData(id));
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("message", "User not found"));
        }
    }

    @GetMapping("/users/{id}/bio")
    public ResponseEntity<?> getUserBioById(@PathVariable UUID id) {
        try {
            if (!hasAccess(AuthUtil.getCurrentUserId(), id)) {
                return ResponseEntity.status(404).body(Map.of("message", "User not found"));
            }
            return ResponseEntity.ok(usersService.getUserBioData(id));
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("message", "User not found"));
        }
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private boolean hasAccess(UUID viewerId, UUID targetId) {
        if (viewerId.equals(targetId)) return true;
        String status = relationshipService.getRelationshipStatus(viewerId, targetId);
        if (!"none".equals(status)) return true;
        List<String> recs = recommendationService.getTopRecommendations(viewerId);
        return recs.contains(targetId.toString());
    }
}